aesd-char-driver/main.c

```diff
@@ -307,14 +317,126 @@ loff_t aesd_llseek(struct file *filp, loff_t offset, int whence)
 
 /* update actual f_pos with new calculated value */
     filp->f_pos = new_f_pos;
-    PDEBUG("\nNew f_pos = %lld", filp->f_pos);
+    PDEBUG("\nUpdated f_pos = %lld", filp->f_pos);
 
 exit:
     mutex_unlock(&dev->mutex);
 
     return new_f_pos;
 }
 
+/**
+ *  @name   aesd_adjust_file_offset
+ *  @brief  adjust *f_pos to the desired entry start offset + desired offset within that entry
+ * 
+ *  @param  filp                pointer to our device
+ *  @param  write_cmd           buffer entry (0-9)
+ *  @param  write_cmd_offset    offset within desired entry
+ * 
+ *  @return 0 on success, error on falure
+*/
+static long aesd_adjust_file_offset(struct file *filp, unsigned int write_cmd, unsigned int write_cmd_offset)
+{
+    struct aesd_dev *dev = filp->private_data;
+    unsigned int command_start_offset = 0;
+    int i = 0;
+    int retval = 0;
+
+    PDEBUG("\naesd_adjust_file_offset");
+    PDEBUG("\n\twrite_cmd: %d, write_cmd_offset: %d", write_cmd, write_cmd_offset);
+
+/* out of range entry (>10) */
+    if(write_cmd > AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED)
+    {
+        retval = -EINVAL;
+        PDEBUG("\nwr_cmd > 10");
+        goto exit;
+    }    
+/* entry not written yet (buffer entry doesnt exist) */
+    if(dev->buffer.entry[write_cmd].buffptr == NULL)
+    {
+        retval = -EINVAL;
+        PDEBUG("\nentry doesn't exist");
+        goto exit;
+    }  
+/* write_cmd_offset is larger than the size of the entry */
+    if(write_cmd_offset >= dev->buffer.entry[write_cmd].size)
+    {
+        retval = -EINVAL;
+        PDEBUG("\noffset larger than size of desired entry");
+        goto exit;
+    }  
+
+/* sum sizes of preceding entries, will be start offset of our desired entry */
+    for(i = 0; i < write_cmd; i++)
+        command_start_offset += dev->buffer.entry[i].size;
+
+/* update *f_pos with entry starting offset + offset within desired entry */
+    filp->f_pos += (command_start_offset + write_cmd_offset);
+    PDEBUG("\nNew *f_pos %lld", filp->f_pos);
+
+exit:
+    return retval;
+}
+
+/**
+ *  @name   aesd_compat_ioctl
+ *  @brief  add ioctl functionality, supports single command
+ * 
+ *  @param  filp    our device
+ *  @param  cmd     ioctl command to execute
+ *  @param  arg     arguments to pass with cmd
+ * 
+ *  @return 0 on success, error on failure
+ * 
+ *  followed general format from https://embetronicx.com/tutorials/linux/device-drivers/ioctl-tutorial-in-linux/
+*/
+long aesd_unlocked_ioctl(struct file *filp, unsigned int cmd, unsigned long arg)
+{
+    long retval = 0;
+    struct aesd_seekto seekto;
+    struct aesd_dev *dev = filp->private_data;
+
+    PDEBUG("\nioctl");
+
+/* obtain mutex, exit on failure */
+    if(mutex_lock_interruptible(&dev->mutex) != 0)
```

Comment:
@mash9756 is this the minimum locking interval?  Is locking needed for `copy_from_user?`